Add historical prices api endpoint

tylerhaugen-stanley opened the pull request.

dblock commented:
Looks good, will merge on green.

dblock commented on lib/iex/endpoints/historial_prices.rb:
```diff
@@ -0,0 +1,21 @@
+      def historical_prices(symbol, options = {})
+        slug = "stock/#{symbol}/chart"
+        slug += "/#{options[:range]}" if options[:range]
+        slug += "/#{options[:date]}" if options[:range] == 'date'
```
This will be wrong if `options[:date]` is missing. I would add a check to `raise ArgumentError` on a missing date when range is date or something like that. Make sure to add specs for the combinations.

tylerhaugen-stanley replied:
Technically this will just return an empty array but that's pretty useless so I'll add the argument error. 

Good call about the specs, pretty sloppy of me. While adding specs I noticed that you can actually receive a different payload if you ask for a specific date and don't include additional query params (ex: chartByDay: true). This will take a bit of time to investigate how to best solve so It'll be a bit before this PR is ready for review again.

dblock replied:
Don't let this block you.

tylerhaugen-stanley replied:
Sounds good. Basically you can get 2 types of data from the IEX endpoint, per day or per min data. The current implementation only supports the per day format.

dblock commented on lib/iex/endpoints/historial_prices.rb:
```diff
@@ -0,0 +1,21 @@
+
+        # We only want options to include query params at this point, remove :range and :date
+        options.delete(:range)
+        options.delete(:date)
```
You're modifying the hash passed in, so either `options = options.dup`, or use `.except` or similar to get a copy without those keys. Let's make sure there are no other places where we did this, I may have missed that.

tylerhaugen-stanley replied:
I didn't see any downside to modifying it but happy to dup it if you prefer. This was the only place I did this.

dblock replied:
The downside is that code like this doesn't do what you'd expect it to do:

```ruby
options = { range: ... }
prices_for_msft = historical_prices('msft', options)
prices_for_goog = historical_prices('goog', options)
```

dblock replied:
Also use `options.key?(:range)`, `nil` is a value and while it doesn't make sense you don't want the caller to rely on you clearing the option, this is more by convention

tylerhaugen-stanley replied:
Yup that's a downside. Thanks

dblock commented on lib/iex/endpoints/historial_prices.rb:
```diff
@@ -0,0 +1,21 @@
+  module Endpoints
+    module HistoricalPrices
+      def historical_prices(symbol, options = {})
+        slug = "stock/#{symbol}/chart"
```
This is a `path` rather than a `slug`, let's call it `path`.

tylerhaugen-stanley commented on lib/iex/resources/resource.rb:
```diff
@@ -14,6 +14,18 @@ def self.float_to_percentage(float_number)
       end
 
+      # Useful for values that are already a percent but we want to convert into a 2 decimal place string
+      def self.percentage_to_string(float_percent)
```
This is essentially a direct copy of `float_to_percentage` but the the values being returned from IEX for historical prices were already a percent just in float format, so the string representation ended up being an incorrect value. Happy to try and combine these two methods with an optional parameter or something different? Don't love the copy paste nature of this.

dblock replied:
I think copy paste is fine I wouldn't worry about it.

tylerhaugen-stanley commented:
I think I've addressed all your concerns @dblock, please let me know if there are more. Also not sure why the travis-ci build failed, I don't see anything in the details page.

dblock commented:
> 1 Error
> 🚫	The TOC found in README.md doesn't match the sections of the file.
> Here's the expected TOC for README.md:
> 
> ```gfm
> # Table of Contents
> 
> - [Installation](#installation)
> - [Usage](#usage)
>   - [Get an API Token](#get-an-api-token)
>   - [Configure](#configure)
>   - [Get a Single Price](#get-a-single-price)
>   - [Get a Quote](#get-a-quote)
>   - [Get a OHLC (Open, High, Low, Close) price](#get-a-ohlc-open-high-low-close-price)
>   - [Get a Market OHLC (Open, High, Low, Close) prices](#get-a-market-ohlc-open-high-low-close-prices)
>   - [Get Historical Prices](#get-historical-prices)
>   - [Get Company Information](#get-company-information)
>   - [Get a Company Logo](#get-a-company-logo)
>   - [Get Recent News](#get-recent-news)
>   - [Get Chart](#get-chart)
>   - [Get Key Stats](#get-key-stats)
>   - [Get Advanced Stats](#get-advanced-stats)
>   - [Get Dividends](#get-dividends)
>   - [Get Earnings](#get-earnings)
>   - [Get Income Statement](#get-income-statement)
>   - [Get Balance Sheet](#get-balance-sheet)
>   - [Get Cash Flow Statement](#get-cash-flow-statement)
>   - [Get Sector Performance](#get-sector-performance)
>   - [Get Largest Trades](#get-largest-trades)
>   - [Get a Quote for Crypto Currencies](#get-a-quote-for-crypto-currencies)
>   - [ISIN Mapping](#isin-mapping)
>   - [Get Symbols](#get-symbols)
>   - [Get List](#get-list)
>   - [Other Requests](#other-requests)
> - [Configuration](#configuration)
>   - [Logging](#logging)
> - [Sandbox Environment](#sandbox-environment)
> - [Errors](#errors)
>   - [SymbolNotFound](#symbolnotfound)
>   - [PermissionDeniedError](#permissiondeniederror)
>   - [ClientError](#clienterror)
> - [Contributing](#contributing)
> - [Copyright and License](#copyright-and-license)
> ```
> 
> Generated by 🚫 [Danger](https://danger.systems/)

The build is failing because of this ^, fix the TOC.

dblock commented on README.md:
```diff
@@ -15,6 +15,7 @@ A Ruby client for the [The IEX Cloud API](https://iexcloud.io/docs/api/).
   - [Get a Quote](#get-a-quote)
   - [Get a OHLC (Open, High, Low, Close) price](#get-a-ohlc-open-high-low-close-price)
   - [Get a Market OHLC (Open, High, Low, Close) prices](#get-a-market-ohlc-open-high-low-close-prices)
+  - [Get Historical Prices](#get-historial-prices)
```
There's a typo in historical, missing `c`.

tylerhaugen-stanley replied:
Ahh I expected Danger to re-post an error after the initial one where I fixed the header but not the tag.

dblock commented on lib/iex/endpoints/historial_prices.rb:
```diff
@@ -0,0 +1,28 @@
+    module HistoricalPrices
+      def historical_prices(symbol, options = {})
+        if options[:range] == 'date'
+          raise ArgumentError if options[:date].nil?
```
`unless options[...]` feels cleaner

tylerhaugen-stanley replied:
Personally I prefer the `if` route but happy to change.

dblock commented on lib/iex/endpoints/historial_prices.rb:
```diff
@@ -0,0 +1,28 @@
+
+        options = options.dup
+        path = "stock/#{symbol}/chart"
+        path += "/#{options[:range]}" if options[:range]
```
`if options.key?(:range)`, which is faster

tylerhaugen-stanley replied:
My bad, misunderstood your previous comment.

dblock commented on lib/iex/endpoints/historial_prices.rb:
```diff
@@ -0,0 +1,28 @@
+        path = "stock/#{symbol}/chart"
+        path += "/#{options[:range]}" if options[:range]
+        # IEX needs dates passed without any formatting, trim out common symbols if passed in.
+        path += "/#{options[:date].tr('-/', '')}" if options[:range] == 'date'
```
This feels wrong. IEX expects a date in a well defined format, not one without dashes or slashes, and it's not for the client to be mangling input. It's a slippery slope when IEX adds support for `-5d` (last 5 days) or something like that and we have to deprecate magical behavior. So either leave the input as is, and document what's required, and/or add support for passing in a DateTime and convert it into the right kind of string.

tylerhaugen-stanley replied:
So i did this because typing in a date as `20201101` is a lot more difficult as a user of the library than typing in `2020-11-01` or `2020/11/01` as most humans would type a date. I like the DateTime approach more through so I'll change that and the documentation.

tylerhaugen-stanley replied:
On second thought, any specific reason for DateTime vs Date? This endpoint has no concept of time in its current implementation.

dblock commented on spec/iex/endpoints/historical_prices_spec.rb:
```diff
@@ -0,0 +1,170 @@
+
+        it 'retrieves historical prices' do
+          expect(subject.size).to eq 23
+          expect(historical_price.date).to eq '2020-10-07'
```
Should this be converted to a `Date` as well?

tylerhaugen-stanley replied:
Yeah.... I think it should be however, all the other endpoints that return dates are also strings. I say lets leave this as is and I'll fix all of them up in a follow up PR.

tylerhaugen-stanley replied:
For this, would you prefer two fields, one for the raw data and one for the converted data? Similar to how dollar values are handled?

dblock replied:
Dollar values are different from dates since those are integers in cents, and formatted values in dollars. I think this should just be a `Date` type. NBD though, I'm going to merge as is and feel free to change it on top before the next release, double-checking that we do it everywhere a date is returned in other instances.

dblock commented on lib/iex/endpoints/historial_prices.rb:
```diff
@@ -0,0 +1,29 @@
+      def historical_prices(symbol, options = {})
+        if options[:range] == 'date'
+          raise ArgumentError unless options[:date].present?
+          raise ArgumentError, 'Date param must be a Date object' unless options[:date].class == Date
```
`.is_a?(Date)` is the proper way to do this, for one it can inherit from `Date`

I think you want to support non-Date input and pass it through in that case. So I would just drop this case. When passing `Date`, call `strftime`, otherwise pass through the value as is.

```ruby
options = options.dup
options[:date] = options[:date].strftime('...') if options[:date].is_a?(Date)
```

tylerhaugen-stanley replied:
Good point about `is_a` thanks.

Fair enough with passing the date through as a string. If the string date is incorrect then the call with simply error out. 👍

dblock reviewed (approved):
I'm happy to merge as is, but consider these two suggestions?

tylerhaugen-stanley commented:
> I'm happy to merge as is, but consider these two suggestions?
> 
> `rubocop -a` will fix the build

Ahh keep forgetting about rubocop. Apologies.
